cmd/mt-parrot/monitor.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"fmt"
+	"github.com/grafana/metrictank/clock"
+	"github.com/grafana/metrictank/stacktest/graphite"
+	"github.com/grafana/metrictank/stats"
+	log "github.com/sirupsen/logrus"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+var (
+	httpError    = stats.NewCounter32("parrot.monitoring.error;error=http")
+	decodeError  = stats.NewCounter32("parrot.monitoring.error;error=decode")
+	invalidError = stats.NewCounter32("parrot.monitoring.error;error=invalid")
+)
+
+var metricsBySeries []partitionMetrics
+
+type seriesStats struct {
+	lastTs           uint32
+	nans             int32   // the partition currently being checked - nope?
+	deltaSum         float64 // sum of abs(value - ts) across the time series
+	numNonMatching   int32   // number of timestamps where value != ts
+	lastSeen         uint32  // the last seen non-NaN time stamp (useful for lag)
+	correctNumPoints bool    // whether the expected number of points were received
+	correctAlignment bool    // whether the last ts matches `now`
+	correctSpacing   bool    // whether all points are sorted and 1 period apart
+}
+
+type partitionMetrics struct {
+	nanCount         *stats.Gauge32 // the number of missing values for each series
+	lag              *stats.Gauge32 // time since the last value was recorded
+	deltaSum         *stats.Gauge32 // the total amount of drift between expected value and actual values
+	nonMatching      *stats.Gauge32 // total number of entries where drift occurred
+	correctNumPoints *stats.Bool    // whether the expected number of points were received
+	correctAlignment *stats.Bool    // whether the last ts matches `now`
+	correctSpacing   *stats.Bool    // whether all points are sorted and 1 period apart
+}
+
+func monitor() {
+	initMetricsBySeries()
+	for tick := range clock.AlignedTickLossless(queryInterval) {
+
+		query := graphite.ExecuteRenderQuery(buildRequest(tick))
+		if query.HTTPErr != nil {
+			httpError.Inc()
+			continue
+		}
+		if query.DecodeErr != nil {
+			decodeError.Inc()
+			continue
+		}
+
+		for _, s := range query.Decoded {
+			processPartitionSeries(s, tick)
+		}
+	}
+}
+
+func processPartitionSeries(s graphite.Series, now time.Time) {
+	partition, err := strconv.Atoi(s.Target)
+	if err != nil {
+		log.Debug("unable to parse partition", err)
+		invalidError.Inc()
+		return
+	}
+	if len(s.Datapoints) < 2 {
+		log.Debugf("partition has invalid number of datapoints: %d", len(s.Datapoints))
+		invalidError.Inc()
+		return
+	}
+
+	serStats := seriesStats{}
+	serStats.lastTs = s.Datapoints[len(s.Datapoints)-1].Ts
+	serStats.correctAlignment = int64(serStats.lastTs) == now.Unix()
```

Comment:
this is not correct. the lastTs should be the highest value that is smaller or equal to now and is divisible by the series interval.

e.g. if series interval is 10, and now is 95 or 99, then lastTs should be 90, but if now is 100 than lastTs should be 100

Reply:
We're querying using an aligned ticker, so `now` should always be evenly divisible, right?

Reply:
no. testMetricsInterval is independent from queryInterval.
queryInterval may be smaller than testMetricsInterval, or consider what happens when queryInterval is 45s and testMetricsInterval is 30s.

I will push some commits that fix all of this